### ISSUE TITLE
Allow cancellation of Process standard IO stream operations on Windows

### ISSUE THREAD

**Issue by madelson** [dotnet/runtime#84290]
### Description

The standard io streams on `Process` are `FileStream`s pointing to the underlying anonymous pipes which on windows do not support async IO (see [this issue](https://github.com/dotnet/runtime/issues/81896)). As a result, canceling hung reads/writes with `CancellationToken` does not work reliably.

This is frustrating because process streams operations are actually fairly likely to hang, either because the buffer fills up for standard input or because the underlying process is hanging/not writing to that stream. It is also frustrating that cancellation works on Linux but not Windows, leading to different behaviors between platforms.

I think this scenario could be supported via the [`CancelSynchronousIo` function in Windows](https://learn.microsoft.com/en-us/windows/win32/fileio/cancelsynchronousio-func); the async IO operations on sync `FileStream`s are async-over-sync so it should be possible to cancel the IO operation on the threadpool thread that is running it.

### Reproduction Steps

```
async Task Main()
	{
	var process = new Process { StartInfo = { } };
	process.StartInfo.FileName = "cmd";
	process.StartInfo.RedirectStandardOutput = true;
	process.StartInfo.RedirectStandardInput = true;
	process.StartInfo.UseShellExecute = false;
	process.Start();
	var t = Task.Run(async () =>
	{
		while (true)
		{
			using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(1));
			if (await process.StandardOutput.BaseStream.ReadAsync(new byte[10], 0, 10, cts.Token) == 0) { break; }
		}
	});
	Console.WriteLine(t.Wait(5000)); // false
	process.Kill();
	await t; // succeeds
}
```

### Expected behavior

`t.Wait(5000);` should throw after 1s when the task becomes canceled.

### Actual behavior

Cancellation never triggers; once the process is killed after 5s the ReadAsync call returns 0 bytes and completes successfully.

### Regression?

No

### Known Workarounds

You can avoid calling `ReadAsync` and instead call `Read` and use `CancelSynchronousIo` yourself.

### Configuration

Windows 10 x64 .NET 6

**Comment by madelson** [comment]
Great to know this approach is already implemented elsewhere. (2) seems appealing because presumably there are other cases where `FileStream` operations can block and `CancellationToken` should just work. Also, I could imagine that people have cast `Process.StandardOutput.BaseStream` to `FileStream` to get the underlying handle for some reason and such code would break if changed (not that that should necessarily prevent the change).

**Comment by adamsitnik** [comment]
> Update FileStream to use the same AsyncOverSyncRead/Write helpers that AnonymousPipeClientStream uses.

Do I understand correctly that it would have an effect of adding a cancellation support for `FileStream(isAsync: false)`? If so, I am supportive of this idea.

**Comment by stephentoub** [comment]
> > Update FileStream to use the same AsyncOverSyncRead/Write helpers that AnonymousPipeClientStream uses.
> 
> Do I understand correctly that it would have an effect of adding a cancellation support for `FileStream(isAsync: false)`? If so, I am supportive of this idea.

Yes

**Comment by stephentoub** [comment]
> Also, I could imagine that people have cast Process.StandardOutput.BaseStream to FileStream to get the underlying handle for some reason and such code would break if changed (not that that should necessarily prevent the change).

While that's possible, a) as you note we still permit such changes, and b) on non-Windows systems that's already a PipeStream rather than a FileStream.

That said, I've addressed this in https://github.com/dotnet/runtime/pull/87103 by changing FileStream to use the same mechanism we used to make PipeStream async-over-sync operations cancelable in .NET 7.  We could subsequently still choose to change Process to use PipeStream on Windows as well, if desired.
